File: python/software/utility.py
from components.software import Software
from interfaces.executable import Executable
from interfaces.upgradable import Upgradable

class Utility(Software, Executable, Upgradable):
    def __init__(self, utility_name="", function="", version="", software_version="", license_key="", size_mb=0.0):
        super().__init__(software_version, license_key, size_mb)
        self._utility_name = utility_name
        self._function = function
        self._version = version

    # Setters
    def set_utility_name(self, utility_name):
        self._utility_name = utility_name

    def set_function(self, function):
        self._function = function

    def set_version(self, version):
        self._version = version

    # Getters
    def get_utility_name(self):
        return self._utility_name

    def get_function(self):
        return self._function

    def get_version(self):
        return self._version

    def execute(self):
        print("Utility executed")

    def terminate(self):
        print("Utility terminated")

    def upgrade(self):
        print("Utility upgraded")
        return True

    def downgrade(self):
        print("Utility downgraded")
        return True

File: cpp/software/Utility.cpp
#pragma once
#include <bits/stdc++.h>

#include "../component/Software.cpp"
#include "../interfaces/Executable.cpp"
#include "../interfaces/Upgradable.cpp"

class Utility : public Software, public Executable, public Upgradable
{
private:
    std::string utilityName;
    std::string function;
    std::string version;

public:
    // Constructor
    Utility() : Software("", "", 0.0), utilityName(""), function(""), version("") {}
    Utility(std::string name, std::string func, std::string ver, std::string softwareVersion, std::string licenseKey, double sizeMB)
        : Software(softwareVersion, licenseKey, sizeMB), utilityName(name), function(func), version(ver) {}

    // Destructor
    ~Utility() {}

    // Setters
    void setUtilityName(const std::string &name)
    {
        utilityName = name;
    }

    void setFunction(const std::string &func)
    {
        function = func;
    }

    void setVersion(const std::string &ver)
    {
        version = ver;
    }

    // Getters
    std::string getUtilityName() const
    {
        return utilityName;
    }

    std::string getFunction() const
    {
        return function;
    }

    std::string getVersion() const
    {
        return version;
    }

    void execute() override
    {
        printf("Utility executed\n");
    }

    void terminate() override
    {
        printf("Utility terminated\n");
    }

    bool upgrade() override
    {
        printf("Utility upgraded\n");
        return true;
    }

    bool downgrade() override
    {
        printf("Utility downgraded\n");
        return true;
    }
};
